node/src/chain/acceptor.rs

```diff
@@ -80,6 +80,8 @@ impl<DB: database::DB, VM: vm::VMExecution, N: Network> Acceptor<N, DB, VM> {
         let mrb_height = mrb.inner().header().height;
         let mrb_state_hash = mrb.inner().header().state_hash;
 
+        let provisioners_list = ContextProvisioners::new(provisioners_list);
```

Comment:
What happens if we reload the node at the beginning of a new epoch? Don't we need the prev-epoch provisioner list?

Reply:
I came to conclusion that there is no need for that
If we start from scratch, it means that the block has been already accepted, then no need to check the winning_certificate of itself.

Does it sounds to you too?

Reply:
We need to recover prev provisioner list here.

Reply:
> I came to conclusion that there is no need for that If we start from scratch, it means that the block has been already accepted, then no need to check the winning_certificate of itself.
> 
> Does it sounds to you too?

Yes it does

Reply:
> We need to recover prev provisioner list here.

Can you explain me why? This is the initial bootstrap phase

Reply:
> I came to conclusion that there is no need for that If we start from scratch, it means that the block has been already accepted, then no need to check the winning_certificate of itself.
> 
> Does it sounds to you too?

Winning_certificate?  Don't we use the prev-epoch/block provisionerst_list to verify `prev_block_cert` ?


```rust
        verify_block_cert(
            prev_block_seed,
            prev_eligible_provisioners,
            mrb.hash,
            mrb.height,
            &new_blk.prev_block_cert,
            mrb.iteration,
            true,
        )
```

Reply:
Prev_block_cert is indeed the winning certificate seen by the prev_block

Reply:
> > I came to conclusion that there is no need for that If we start from scratch, it means that the block has been already accepted, then no need to check the winning_certificate of itself.
> > Does it sounds to you too?
> 
> Winning_certificate? Don't we use the prev-epoch/block provisionerst_list to verify `prev_block_cert` ?
> 
> ```rust
>         verify_block_cert(
>             prev_block_seed,
>             prev_eligible_provisioners,
>             mrb.hash,
>             mrb.height,
>             &new_blk.prev_block_cert,
>             mrb.iteration,
>             true,
>         )
> ```

I think his point is that if we accepted the block we already verified `prev_block_cert` so there's no need to do it again.

Reply:
Genesis- NoTxs
Block-1 - NoTxs  (prev_provisioner list)
Block-2 - Unstake Txs (current list, after one unstake)

> Node is restarted.
> Node loads current_list only on bootstrap:

Trying to verify Block-3, prev_block_cert but knows only `(current list)`

Reply:
I see your point. I think you're right

Reply:
oh gotcha, you're right.

Let me see if I can arrange it, probably I'll need to store it on db, because it's not possible to query the vm for the previous state.

Reply:
Also, does this issue also apply to `update_most_recent_block`? If we revert to the last final block, and then accept a new block, we need to be able to verify the previous block certificate

Reply:
> update_most_recent_block

Um, I think you're absolute right here. If we fallback to a block that has `unstake` tx in it, we cannot accept another block anymore.

Reply:
Ideally, if we have these two tasks doable:

- Apply unstake transaction at epoch block only
- Create snapshot of last epoch block in VM

We can always fetch on demand both `current` and `previous` provisioners_lists from VM.

Reply:
I simply changed the vm to be able to ask for the list of provisioners up to the block preceding the last finalized

That should cover all the use cases